File: server/flow/report.py
"""测试运行报告：断言结果 + 节点错误 + 失败证据(截图) + 日志；可导出 JSON/JUnit。"""
from __future__ import annotations

import time
from xml.sax.saxutils import escape, quoteattr

from ..i18n import tr


class Report:
    def __init__(self):
        self.started = time.time()
        self.finished = None
        self.asserts: list[dict] = []
        self.errors: list[dict] = []
        self.logs: list[str] = []
        self.error: str | None = None

    def add_assert(self, ok: bool, message: str, node_id, evidence: str | None = None) -> None:
        self.asserts.append({
            "ok": bool(ok), "message": message or "", "node": node_id,
            "evidence": evidence,
        })

    def add_error(self, node_id, message: str, evidence: str | None = None) -> None:
        self.errors.append({"node": node_id, "message": message, "evidence": evidence})

    def log(self, text: str) -> None:
        self.logs.append(str(text))

    def finalize(self, error: str | None = None) -> None:
        self.finished = time.time()
        self.error = error

    @property
    def duration(self) -> float:
        return round((self.finished or time.time()) - self.started, 3)

    @property
    def passed(self) -> bool:
        return self.error is None and not self.errors and all(a["ok"] for a in self.asserts)

    def to_dict(self) -> dict:
        return {
            "passed": self.passed,
            "error": self.error,
            "duration": self.duration,
            "asserts": self.asserts,
            "errors": self.errors,
            "total": len(self.asserts),
            "failed": sum(1 for a in self.asserts if not a["ok"]),
            "logs": self.logs,
        }

    def to_junit(self, suite_name: str = "flow", lang: str = "zh") -> str:
        n = len(self.asserts)
        failures = sum(1 for a in self.asserts if not a["ok"])
        errors = len(self.errors)
        out = [
            '<?xml version="1.0" encoding="UTF-8"?>',
            f'<testsuite name={quoteattr(suite_name)} tests="{n}" '
            f'failures="{failures}" errors="{errors}" time="{self.duration}">',
        ]
        for a in self.asserts:
            name = a["message"] or f'assert@node{a["node"]}'
            out.append(f'  <testcase classname="node{a["node"]}" name={quoteattr(name)}>')
            if not a["ok"]:
                detail = tr(lang, "report.assert_failed", "断言失败")
                if a.get("evidence"):
                    detail += f"\n{tr(lang, 'report.evidence', '证据')}: {a['evidence']}"
                out.append(f'    <failure message={quoteattr(name)}>{escape(detail)}</failure>')
            out.append("  </testcase>")
        for e in self.errors:
            out.append(f'  <testcase classname="node{e["node"]}" name="error">')
            detail = e["message"] + (
                f"\n{tr(lang, 'report.evidence', '证据')}: {e['evidence']}" if e.get("evidence") else "")
            out.append(f'    <error message={quoteattr(e["message"][:120])}>{escape(detail)}</error>')
            out.append("  </testcase>")
        if self.error and not self.errors:
            out.append('  <testcase classname="run" name="run">')
            out.append(f'    <error message={quoteattr(self.error)}>{escape(self.error)}</error>')
            out.append("  </testcase>")
        out.append("</testsuite>")
        return "\n".join(out)

File: internal/flow/report.go
package flow

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

type AssertResult struct {
	OK       bool   `json:"ok"`
	Message  string `json:"message"`
	Node     int64  `json:"node"`
	Evidence string `json:"evidence,omitempty"`
}

type NodeErrorResult struct {
	Node     int64  `json:"node"`
	Message  string `json:"message"`
	Evidence string `json:"evidence,omitempty"`
}

type Report struct {
	Started  time.Time         `json:"-"`
	Finished time.Time         `json:"-"`
	Passed   bool              `json:"passed"`
	Error    string            `json:"error,omitempty"`
	Duration float64           `json:"duration"`
	Assert   []AssertResult    `json:"asserts"`
	Errors   []NodeErrorResult `json:"errors"`
	Total    int               `json:"total"`
	Failed   int               `json:"failed"`
	Logs     []string          `json:"logs"`
}

func NewReport() *Report {
	return &Report{Started: time.Now()}
}

func (r *Report) AddAssert(ok bool, message string, nodeID int64, evidence string) {
	r.Assert = append(r.Assert, AssertResult{OK: ok, Message: message, Node: nodeID, Evidence: evidence})
}

func (r *Report) AddError(nodeID int64, message string) {
	r.Errors = append(r.Errors, NodeErrorResult{Node: nodeID, Message: message})
}

func (r *Report) AddLog(message string) {
	r.Logs = append(r.Logs, message)
}

func (r *Report) Finalize(err error) {
	r.Finished = time.Now()
	r.Duration = float64(r.Finished.Sub(r.Started).Milliseconds()) / 1000
	if err != nil {
		r.Error = err.Error()
	}
	r.Total = len(r.Assert)
	r.Failed = 0
	for _, item := range r.Assert {
		if !item.OK {
			r.Failed++
		}
	}
	r.Passed = r.Error == "" && len(r.Errors) == 0 && r.Failed == 0
}

func ReportFromEvents(events []Event, err error) *Report {
	report := NewReport()
	for _, event := range events {
		switch event.Type {
		case "assert":
			report.AddAssert(event.Status == "ok", event.Info, event.ID, event.URL)
		case "alert":
			report.AddLog(event.Message)
		}
	}
	var nodeErr NodeError
	if errors.As(err, &nodeErr) {
		report.AddError(nodeErr.NodeID, nodeErr.Error())
	}
	report.Finalize(err)
	return report
}

func (r *Report) ToJUnit(suiteName string) string {
	if suiteName == "" {
		suiteName = "flow"
	}
	var out []string
	out = append(out, `<?xml version="1.0" encoding="UTF-8"?>`)
	out = append(out, fmt.Sprintf(
		`<testsuite name=%q tests="%d" failures="%d" errors="%d" time="%.3f">`,
		xmlAttr(suiteName), r.Total, r.Failed, len(r.Errors), r.Duration,
	))
	for _, item := range r.Assert {
		name := item.Message
		if name == "" {
			name = fmt.Sprintf("assert@node%d", item.Node)
		}
		out = append(out, fmt.Sprintf(`  <testcase classname="node%d" name=%q>`, item.Node, xmlAttr(name)))
		if !item.OK {
			detail := "assert failed"
			if item.Evidence != "" {
				detail += "\nevidence: " + item.Evidence
			}
			out = append(out, fmt.Sprintf(`    <failure message=%q>%s</failure>`, xmlAttr(name), xmlText(detail)))
		}
		out = append(out, "  </testcase>")
	}
	for _, item := range r.Errors {
		out = append(out, fmt.Sprintf(`  <testcase classname="node%d" name="error">`, item.Node))
		detail := item.Message
		if item.Evidence != "" {
			detail += "\nevidence: " + item.Evidence
		}
		out = append(out, fmt.Sprintf(`    <error message=%q>%s</error>`, xmlAttr(shortString(item.Message, 120)), xmlText(detail)))
		out = append(out, "  </testcase>")
	}
	if r.Error != "" && len(r.Errors) == 0 {
		out = append(out, `  <testcase classname="run" name="run">`)
		out = append(out, fmt.Sprintf(`    <error message=%q>%s</error>`, xmlAttr(r.Error), xmlText(r.Error)))
		out = append(out, "  </testcase>")
	}
	out = append(out, "</testsuite>")
	return strings.Join(out, "\n")
}

func xmlText(value string) string {
	return strings.NewReplacer(
		"&", "&amp;",
		"<", "&lt;",
		">", "&gt;",
	).Replace(value)
}

func xmlAttr(value string) string {
	return strings.NewReplacer(
		"&", "&amp;",
		"<", "&lt;",
		">", "&gt;",
		`"`, "&quot;",
		"'", "&apos;",
	).Replace(value)
}

func shortString(value string, max int) string {
	if max <= 0 || len(value) <= max {
		return value
	}
	return value[:max]
}
